Transient failure

batiati opened the pull request:
This PR introduces the new concept of "transient failure", which are events that fail due to the system's status rather than the request data. The list of transient errors is as follows:

create_accounts: none
create_transfers: `debit_account_not_found`, `credit_account_not_found`, `pending_transfer_not_found`, `exceeds_credits`, `exceeds_debits`, `debit_account_already_closed`, and `credit_account_already_closed`.

With this, we introduce the invariant: **Once a transfer fails with a transient error code, it must always fail with `id_already_failed` when retried with the same ID.**
This ensures that a user will never encounter a "transient error code" twice for the same ID, providing the application with stronger guarantees that the outcome of a transfer will remain consistent across retries.

Additionally, this changes the precedence of the idempotency checks (i.e., all `exists_*` checks) to the beginning, right after `id_must_not_be_zero` and `id_must_not_be_int_max`.

The objectless `id` is stored in the `IdTree` with a zeroed `timestamp`.

The Groove is not aware of "transient failures" but instead it has a configuration that enables "orphaned IDs", which are IDs stored without a corresponding object.

We are using the ObjectCache (with an empty object) to cache such IDs, which can be beneficial for pathological workloads that retry the same failed event in a loop.

batiati committed:
[StateMachine] Transient failure
This introduces the new concept of "transient failure," which are events that fail due to the system's status rather than the request data. The list of transient errors is as follows:

create_accounts: none
create_transfers: `debit_account_not_found`, `credit_account_not_found`, `pending_transfer_not_found`, `exceeds_credits`, `exceeds_debits`, `debit_account_already_closed, and `credit_account_already_closed`.

With this, we introduce the invariant: "Once a transfer fails with a transient error code, it must always fail with `id_already_failed` when retried with the same ID." This ensures that a user will never encounter a "transient error code" twice for the same ID, providing the application with stronger guarantees that the outcome of a transfer will remain consistent across retries.

Additionally, this changes the precedence of the idempotency checks (i.e., all `exists_*` checks) to the beginning, right after `id_must_not_be_zero` and `id_must_not_be_int_max`.

The objectless `id` is stored in the `IdTree` with a zeroed `timestamp`.

The Groove is not aware of "transient failures" but instead it has a configuration that enables "orphaned IDs", which are IDs stored without a corresponding object.
We are using the ObjectCache (with an empty object) to cache such IDs, which can be beneficial for pathological workloads that retry the same failed event in a loop.

batiati committed:
[Clients] Transient failure

matklad commented:
>Transient failure 

Upon receiving an email, got spooked that our CI went flaky!

batiati committed:
Tidy: line lenght

batiati commented on src/tigerbeetle.zig:
```diff
@@ -286,31 +298,65 @@ pub const CreateTransferResult = enum(u32) {
+                .name = @tagName(result),
+                .value = index,
+            };
+        }
```
It started to become a pain to keep the enum artificially ordered. Now, we declare the values in precedence order and generate an ordered enum at compile time to use in the `EnumSet` as a workaround.

sentientwaffle replied:
Could this instead `sort` the original enum's fields, so the enum values still correspond to the original enum's? Since the precedence ordering is really just "documentation", we could even do that for the original enum, e.g.

```
const CreateTransferResult = EnumSorted(enum { ... })
```

batiati replied:
That's a nice idea.

> Since the precedence ordering is really just "documentation"

One side effect of this change that I liked was that the client bindings are now also generated according to precedence. It can be quite helpful when reading the code (in other languages) because it matches the docs.

sentientwaffle replied:
Good point; I completely forgot about how it would impact the client implementation!

sentientwaffle commented on src/tigerbeetle.zig:
```diff
@@ -286,31 +298,65 @@ pub const CreateTransferResult = enum(u32) {
+    debit_account_already_closed = 65,
+    credit_account_already_closed = 66,
+
+    // Last item: id_already_failed = 68.
```
:thinking: Why is this listed here? I thought that its position after `exists_*` was correct.

batiati replied:
This comment was intended to be a hint of the last element.
Just to help the next person who will add a new element to follow the numeric order without having to manually browse all values.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -1681,6 +1687,51 @@ pub fn StateMachineType(
+                        .credit_account_already_closed,
+                        => true,
+                        else => false,
+                    };
```
What do you think of defining a new `CreateTransferResult` function `fn transient(CreateTransferResult) bool`, which uses an exhaustive switch? Listing all of the results is quite verbose, but it would make it hard to add a new result code without thinking through whether it should be transient.

sentientwaffle reviewed (commented):
I like this approach! Handling `exists` as early as possible makes a lot of sense even in ignoring `id_already_failed`!

batiati committed:
[StateMachine] Review
- Upgrade strategy for old clients.
- Consolidate checks into a single `transient` function.

batiati committed:
Add vopr tests
-Workload has a list of IDs that failed with transient codes and retries them to confirm the `id_already_failed` result.
- Nit: Rename `Self` to the type's name.

batiati committed:
Generate client bindings

batiati committed:
[StateMachine] Fix error code precedence.

batiati committed:
[Docs]
- Reorder error code precedence.
- Introduce `transient errors`.
- Fix duplicated `debit_account_already_closed`.

batiati committed:
Generate client bindings

batiati committed:
[Scan] Support the timestamp=0 case.

sentientwaffle commented on src/lsm/groove.zig:
```diff
@@ -820,10 +843,25 @@ pub fn GrooveType(
+
+                    if (groove_options.orphaned_ids and
+                        id_tree_value.timestamp == 0)
+                    {
```
Could this be:
```suggestion
                    if (id_tree_value.timestamp == 0) {
                        assert(groove_options.orphaned_ids);
```
(And likewise in `lookup_id_callback` too.)

sentientwaffle commented on src/tigerbeetle.zig:
```diff
@@ -286,31 +300,143 @@ pub const CreateTransferResult = enum(u32) {
+        assert(set.count() == set.capacity());
+    }
+
+    /// TODO(zig): This enum is be ordered by precedence, but it crashes
```
Typo "is be"

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -1855,6 +1890,12 @@ pub fn StateMachineType(
+                .found => |e| return create_account_exists(a, e),
+                .orphaned_id => unreachable,
+                .not_found => {},
+            }
```
This is equivalent to
```
            if (self.forest.grooves.accounts.get(a.id)) |e| {
                return create_account_exists(a, e);
            }
```

batiati replied:
Although it does not apply to Accounts, I still prefer to explicitly handle `orphaned_id => unreachable` both for clarity and symmetry with the idempotency validation in `create_transfer`.

The helper functions `self.get_transfer(id)` and `self.get_account(id)` should be used in situations where the "orphaned_id" case is not a concern and can even be interpreted as "not found" (e.g. for pending transfers).

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -4535,9 +4667,16 @@ test "create_transfers: per-transfer balance invariant" {
-        \\ transfer T1 A1 A2 40   _  _  _  _    0 L1 C1 LNK   _ _   _   _   _ _ _ _ _ _ ok
-        \\ transfer T2 A2 A3  1   _  _  _  _    0 L1 C1 LNK PEN _   _ BDR   _ _ _ _ _ _ ok
-        \\ transfer T3 A2 A3  0  T2  _  _  _    0 L1 C1   _   _ _ VOI   _   _ _ _ _ _ _ ok
+        // Ids failed in a linked chain can be reused, buy
```
```suggestion
        // Ids failed in a linked chain can be reused, but
```

sentientwaffle commented on docs/reference/requests/create_transfers.md:
```diff
@@ -259,11 +381,21 @@ The transfer was not created. [`Transfer.code`](../transfer.md#code) is zero, bu
 The transfer was not created. [`Transfer.debit_account_id`](../transfer.md#debit_account_id) must
 refer to an existing `Account`.
 
+This is a [`transient error`](#id_already_failed).
```
I don't think "transient error" should be in backticks, since it isn't an identifier.

sentientwaffle commented on src/lsm/scan_builder.zig:
```diff
@@ -441,6 +441,14 @@ pub fn ScanType(
+                        if (index == .id) {
+                            // When iterating over `IdTree` it can return a timestamp zero, which
+                            // indicates an orphaned id.
+                            if (Groove.config.orphaned_ids and value.timestamp == 0) continue;
```
Could this be:
```suggestion
                            if (value.timestamp == 0) {
                                assert(Groove.config.orphaned_ids);
                                continue;
                            }
```

sentientwaffle commented on src/state_machine/workload.zig:
```diff
@@ -948,6 +1004,28 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+                    self.transient_failures.putNoClobber(
+                        transfers[item.index].id,
+                        {},
+                    ) catch unreachable;
```
Alternatively, if the hash map had some (arbitrary, fixed, constant) capacity, then this could insert iff it is not full.

sentientwaffle reviewed (commented):
Note that cfo found a forest fuzzer failure too.

batiati committed:
Fix fuzz lsm_forest
SEED: ./zig/zig build fuzz -- lsm_forest 5253265565389130585

batiati committed:
Fix typo

batiati committed:
Review
- Clarify assertions.
- Static limit for VOPR's transient failures.

batiati committed:
[Docs] Review.

batiati committed:
Fix fuzzer lsm_forest.
SEED: ./zig/zig build -Drelease fuzz -- lsm_forest 13382701409420583093

sentientwaffle commented on src/tigerbeetle.zig:
```diff
@@ -400,8 +400,8 @@ pub const CreateTransferResult = enum(u32) {
         assert(set.count() == set.capacity());
     }
 
-    /// TODO(zig): This enum is be ordered by precedence, but it crashes
```
s/"This enum"/CreateTransferResult to clarify that we aren't referring to `Ordered`.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -2524,7 +2599,18 @@ pub fn StateMachineType(
+            return switch (self.forest.grooves.accounts.get(id)) {
+                .found => |a| a,
+                .not_found => null,
+                .orphaned_id => unreachable,
```
Alternatively,
```suggestion
                .found_object => |a| a,
                .found_orphaned_id => unreachable,
                .not_found => null,
```

batiati commented:
I almost forgot to mention: Imported events are **not** handled as a special case.
After all, it feels like transient errors should take precedence over importing legacy events, as they are part of the financial consistency checks.

What are your thoughts @sentientwaffle?

batiati committed:
Review
- Rename `get` result to `found_object` and `found_orphaned_id`.
- Clarify the `CreateTransferResult.Ordered` enum documentation.

sentientwaffle commented:
> After all, it feels like transient errors should take precedence over importing legacy events, as they are part of the financial consistency checks.

I don't think we need to special-case transient errors + import events -- I think that the import script should use a single tigerbeetle client, not a stateless tigerbeetle api server, so transient errors shouldn't happen during imports anyway.

sentientwaffle commented on src/state_machine/workload.zig:
```diff
@@ -453,11 +475,18 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+                transfers,
+                results,
+            );
+            if (failed_ids_count > 0) return failed_ids_count;
```
:thinking: I'd rather we allowed mixed failed-ids-and-regular-transfer batches. Otherwise I'm worried that every other batch will be a failed transfers batch, which would change the coverage of LSM (i.e. every other batch wouldn't be inserting objects).

Instead of removing the id (`self.transient_failures.popOrNull()`), change `transient_failures` to an arraylist (don't worry about duplicates) and select a random element (but don't remove it). Then when we get back `id_already_failed` we can verify that it is in the list.

sentientwaffle commented:
When (or before) 0.16.4 is released, can you add a comment to our [breaking changes tracking issue](https://github.com/tigerbeetle/tigerbeetle/issues/2231) explaining transient errors?

batiati committed:
Revamp the VOPR logic to insert retries in the middle of the batch.

batiati committed:
Review: Prevent retrying within a linked chain and reduce the probability of retries.
